File: p4/Deepspace/NumericDamage.rb
require_relative 'Damage'
require_relative 'NumericDamageToUI'

module Deepspace
	class NumericDamage < Damage
	
		public_class_method :new
		
		def initialize(w,s)
			super(s)
			@nWeapons = w
		end
		
		attr_reader:nWeapons
		
		def adjust(w,s)
		    if (s.count > @nShields)
		        new_nShields = @nShields
		    else
		        new_nShields = s.count
		    end
		    
		    if w.count > @nWeapons
		        new_nWeapons = @nWeapons
		    else
		        new_nWeapons = w.count
		    end     
		    NumericDamage.new(new_nWeapons,new_nShields)
		    
		end 
		
		
		def discardWeapon(w)        	
		    	if @nWeapons-1 >= 0 
		        	@nWeapons = @nWeapons-1
		    	end 
		end 
		
		
		def hasNoEffect()          
			return @nShields + @nWeapons == 0          
		end
		
		def getUIversion()
			Deepspace::NumericDamageToUI.new(self)
		end 
	
	end
end

File: p5/deepspace/src/deepspace/NumericDamage.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package deepspace;

import java.util.ArrayList;
/**
 *
 * @author cristobalmer
 */
public class NumericDamage extends Damage{
    
    private int nWeapons;
    
    public NumericDamage(int _nWeapons, int _nShields) {
        super(_nShields);
        nWeapons = _nWeapons;
    }
    
    public NumericDamage(NumericDamage other){
        this(other.nWeapons, other.getNShields());
    }
    
    @Override
    public NumericDamage copy(){
        return new NumericDamage(nWeapons, getNShields());
    }
    
    public int getNWeapons(){
        return nWeapons;
    }
    
    @Override
    public boolean hasNoEffect(){
        return getNShields() + nWeapons == 0;
    }
    
    @Override
    public NumericDamage adjust(ArrayList<Weapon> w, ArrayList<ShieldBooster> s){
        int shields = Integer.min(s.size(), getNShields());
        return new NumericDamage(Integer.min(nWeapons, w.size()), shields);
    }
    
    @Override
    public void discardWeapon(Weapon w){
        if (nWeapons > 0)
            nWeapons--;
    }
    
    @Override
    public String toString() {
        String message = "[NumericDamage] -> "
                + "Number of shields: " + getNShields()
                + ", Number of weapons: " + nWeapons;
        return message;
    }
    
    @Override
    public NumericDamageToUI getUIversion() {
        return new NumericDamageToUI(this);
    }
}
